### PR TITLE
QuotesBear: Add force_preferred_quotation option

### PR DESCRIPTION
**Pull request by newbazz**
Add a param to force the preffered quotation.

Closes https://github.com/coala/coala-bears/issues/1005

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by newbazz** [comment]
This is a rough PR and i need some help on this.

File: bears/general/QuotesBear.py [hunk 1]
```diff
@@ -72,11 +75,14 @@ def run(self, filename, file, dependency_results,
 
         ranges = dependency_results[AnnotationBear.name][0].contents['strings']
 
-        for string_range in ranges:
-            if (file[string_range.start.line-1][string_range.start.column-1] ==
-                    preferred_quotation):
-                continue
+        if not force_preferred_quotation:
+            for string_range in ranges:
+                if (file[string_range.start.line-1][string_range.start.column-1]
```
[the remark this answers is not in the document]
Comment (newbazz) [reply bears/general/QuotesBear.py]: How to solve this?

[the remark this answers is not in the document]
Comment (mridubhatnagar) [reply bears/general/QuotesBear.py]: if (file[string_range.start.line-1]
[string_range.start.column-1] 
== preferred_quotation):

should work. It seems. But, yes will lead to poor readability.

[the remark this answers is not in the document]
Comment (RaiVaibhav) [reply bears/general/QuotesBear.py]: Define it into some other variable with a suitable name, it will safe both :)
or use `Ignore`, only as a last option

**Comment by newbazz** [comment]
@Makman2 please review

**Comment by newbazz** [comment]
@Makman2 review ;)

File: bears/general/QuotesBear.py [hunk 2]
```diff
@@ -73,10 +76,13 @@ def run(self, filename, file, dependency_results,
         ranges = dependency_results[AnnotationBear.name][0].contents['strings']
 
         for string_range in ranges:
-            if (file[string_range.start.line-1][string_range.start.column-1] ==
+            if (file[string_range.start.line-1][
+                    string_range.start.column-1] ==
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: unnecessary change

File: bears/general/QuotesBear.py [hunk 3]
```diff
@@ -73,10 +76,13 @@ def run(self, filename, file, dependency_results,
...
                     preferred_quotation):
                 continue
 
-            if string_range.start.line == string_range.end.line:
+            if (string_range.start.line ==
+                    string_range.end.line):
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: unnecessary change

File: tests/general/QuotesBearTest.py [hunk 4]
```diff
@@ -66,34 +66,32 @@ def test_valid_quotes(self):
                           dependency_results=self.dep_results) as results:
             self.assertEqual(len(results), 0)
 
+        self.uut.run('F', self.single_quote_file, self.dep_results, "'", True)
+
+        self.uut.run('F', self.single_quote_file, self.dep_results, "'", False)
```
Comment (Makman2) [inline tests/general/QuotesBearTest.py]: without an assertion this has no test value^^

File: tests/general/QuotesBearTest.py [hunk 5]
```diff
@@ -66,34 +66,32 @@ def test_valid_quotes(self):
...
-                             '--- \n'
-                             '+++ \n'
-                             '@@ -2,5 +2,5 @@\n'
-                             " '''\n"
-                             ' Multiline string\n'
-                             " '''\n"
-                             '-"a string with double quotes!"\n'
-                             "+'a string with double quotes!'\n"
-                             " 'A single quoted string with \" in it'\n")
+            self.assertEqual(res_list[0].diffs[self.filename]._file,
```
Comment (Makman2) [inline tests/general/QuotesBearTest.py]: hmm not necessary to change, also `_file` is not part of the public API, what makes tests implementation-detail-dependent. Please use the old `unified_diff` attribute.
Same above.

**Comment by Makman2** [comment]
unack f1d8736

File: tests/general/QuotesBearTest.py (hunk 5 above)
Comment (newbazz) [reply tests/general/QuotesBearTest.py]: `unified_diff` theres no attribute named this. This is what the error shows up.And also `pprint(vars())` also shows no attribute named `unified_diff`.

**Comment by newbazz** [comment]
changes made ;)

File: tests/general/QuotesBearTest.py (hunk 5 above)
Comment (Makman2) [reply tests/general/QuotesBearTest.py]: > `res_list[0].diffs[self.filename].unified_diff`

This was written before :O

**Comment by newbazz** [comment]
@Makman2 changes done :P

**Comment by Makman2** [comment]
Where are tests? :O

**Comment by Makman2** [comment]
unack 9cec2bd

**Comment by newbazz** [comment]
@Makman2 changes done ;P

File: tests/general/QuotesBearTest.py [hunk 6]
```diff
@@ -97,3 +99,21 @@ def test_invalid_quotes(self):
                              '-"a string with double quotes!"\n'
                              "+'a string with double quotes!'\n"
                              " 'A single quoted string with \" in it'\n")
+
+        self.section['preferred_quotation'] = '"'
+        self.force_preferred_quotation = True
+        with execute_bear(self.uut, self.filename, self.double_quote_file,
+                          dependency_results=self.dep_results,
+                          force_preferred_quotation=self.
+                          force_preferred_quotation) as results:
```
Comment (Makman2) [inline tests/general/QuotesBearTest.py]: please also test if `force_preferred_quotation=True`, I mean that's what you intend to test, right?

Comment (Makman2) [reply tests/general/QuotesBearTest.py]: (also not necessary imo to put this into a test-wide variable, just do inline `False`^^)

File: bears/general/QuotesBear.py [hunk 7]
```diff
@@ -51,7 +53,8 @@ def correct_single_line_str(self, filename, file, sourcerange,
                      diff.affected_code(filename), diffs={filename: diff})
 
     def run(self, filename, file, dependency_results,
-            preferred_quotation: str='"'):
+            preferred_quotation: str='"',
+            force_preferred_quotation: bool=False):
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: actually we already have `preferred_quotation`, what does `force_preferred_quotation` now do?

Comment (Makman2) [reply bears/general/QuotesBear.py]: nvm

File: tests/general/QuotesBearTest.py (hunk 6 above)
Comment (newbazz) [reply tests/general/QuotesBearTest.py]: It is good to have one right?

**Comment by newbazz** [comment]
ping @Makman2  ;)

File: tests/general/QuotesBearTest.py [hunk 8]
```diff
@@ -97,3 +97,19 @@ def test_invalid_quotes(self):
...
+            self.assertEqual(len(res_list), 1)
+            self.assertEqual(res_list[0].diffs[self.filename].unified_diff,
+                             '--- \n'
+                             '+++ \n'
+                             '@@ -3,4 +3,4 @@\n'
+                             ' Multiline string\n'
+                             " '''\n"
+                             ' "a string with double quotes!"\n'
+                             "-'A single quoted string with \" in it'\n"
+                             '+"A single quoted string with \" in it"\n')
```
Comment (Makman2) [inline tests/general/QuotesBearTest.py]: Shouldn't it be `\\"`, because the escape character has to appear in the patch?

Comment (newbazz) [reply tests/general/QuotesBearTest.py]: no, why the escape character needs to appear in the patch?

Comment (newbazz) [reply tests/general/QuotesBearTest.py]: no need right? @Makman2

Comment (Makman2) [reply tests/general/QuotesBearTest.py]: Let's consider the test file:
```
'''
Multiline string
'''
"a string with double quotes!"
'A single quoted string with " in it'
```
Because we force quotation with `"`, the last string has to be changed because it uses `'` quotes (let's not care for `'''` right now):
```
'A single quoted string with " in it'
```
The single quotes need to be replaced with double quotes:
```
"A single quoted string with " in it"
```
The problem: The double quote in the middle. It's part of the string, so it has to be escaped:
```
"A single quoted string with \" in it"
```
Because you are entering the string here in Python, you have to escape the `\`, otherwise it just vanishes (just copy the string here into a python interactive shell and print it again, the backslash is gone, because it's interpreted).
So the final string here should be:
```
'+"A single quoted string with \\" in it"\n'
```
This produces then valid code when e.g. used for python strings. But this patch right now would produce invalid code, python would probably say something like "`it` is not defined" (because it sees `"A single quoted string with " in it ...`)

Comment (newbazz) [reply tests/general/QuotesBearTest.py]: Ohk so can you help me how to approach.

**Comment by newbazz** [comment]
@Makman2 changes made ;)

File: bears/general/QuotesBear.py [hunk 9]
```diff
@@ -34,9 +35,16 @@ def correct_single_line_str(self, filename, file, sourcerange,
...
-        if preferred_quotation in str_contents:
+        if (preferred_quotation in str_contents and
+                not force_preferred_quotation):
             return
 
+        elif preferred_quotation in str_contents:
+            str_split = str_contents.split(preferred_quotation)
+            for i in range(len(str_split) - 1):
+                str_split[i] += '\\'
+            str_contents = preferred_quotation.join(str_split)
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: maybe use `str.replace`? Then you don't need the `if` probably, just replace any of those characters if available :+1:

Comment (newbazz) [reply bears/general/QuotesBear.py]: Yeah str.replace can be used, but then also the `if` is to be used.

**Comment by newbazz** [comment]
@Makman2 changed made ;P

File: bears/general/QuotesBear.py [hunk 10]
```diff
@@ -34,9 +35,15 @@ def correct_single_line_str(self, filename, file, sourcerange,
...
 
-        if preferred_quotation in str_contents:
+        if (preferred_quotation in str_contents and
+                not force_preferred_quotation):
             return
 
+        elif preferred_quotation in str_contents:
+            add_escape = '\\'+preferred_quotation
+            str_contents = str_contents.replace(preferred_quotation,
+                                                add_escape)
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: I think you don't need the if, `replace` won't replace if there are no `preferred_quotation` inside `str_contents` ;)

Also I would inline `add_escape`:
```
str_contents = str_contents.replace(preferred_quotation, '\\' + preferred_quotation)
```
or you could actually use `coala_utils.string_process.escape` :) But this is optional, though this function is designed for such scenarios :)

Another thing: It may be worth to add a comment here explaining what this snippet does: Escaping quotation inside the string that would conflict with the preferred quotation used to surround the string :+1:

Comment (newbazz) [reply bears/general/QuotesBear.py]: Yeah about the comments i was thinking it to make a newcomer issue ;P

Comment (newbazz) [reply bears/general/QuotesBear.py]: and making add_escape inline as in `str_contents.replace(preferred_quotation,add_escape)` ?

**Comment by Makman2** [comment]
unack 92931f5

**Comment by newbazz** [comment]
@Makman2 changes made ;P

File: bears/general/QuotesBear.py [hunk 11]
```diff
@@ -34,9 +35,16 @@ def correct_single_line_str(self, filename, file, sourcerange,
         str_contents = file[sourcerange.start.line - 1][
                        sourcerange.start.column:sourcerange.end.column-1]
 
-        if preferred_quotation in str_contents:
+        if (preferred_quotation in str_contents and
+                not force_preferred_quotation):
             return
 
+        """
+        Escape the preffered quotes if its already present.
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: --> Escape preferred quotes if present in string.

File: bears/general/QuotesBear.py [hunk 12]
```diff
@@ -34,9 +35,16 @@ def correct_single_line_str(self, filename, file, sourcerange,
...
+        """
+        add_escape = '\\'+preferred_quotation
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: Optional: I prefer spaces around operators :+1:

Comment (Makman2) [reply bears/general/QuotesBear.py]: Anyway, I would definitely inline this, the variable name `add_escape` is bad :3

File: tests/general/QuotesBearTest.py [hunk 13]
```diff
@@ -97,3 +97,19 @@ def test_invalid_quotes(self):
...
+            self.assertEqual(len(res_list), 1)
+            self.assertEqual(res_list[0].diffs[self.filename].unified_diff,
+                             '--- \n'
+                             '+++ \n'
+                             '@@ -3,4 +3,4 @@\n'
+                             ' Multiline string\n'
+                             " '''\n"
+                             ' "a string with double quotes!"\n'
+                             "-'A single quoted string with \" in it'\n"
+                             '+"A single quoted string with \\" in it"\n')
```
Comment (Makman2) [inline tests/general/QuotesBearTest.py]: could you also test a case with `'` as preferred quotes, and a string where no preferred quotes are inside? :+1:

**Comment by newbazz** [comment]
@Makman2 done :D

File: bears/general/QuotesBear.py [hunk 14]
```diff
@@ -30,13 +31,22 @@ def correct_single_line_str(self, filename, file, sourcerange,
...
         str_contents = file[sourcerange.start.line - 1][
                        sourcerange.start.column:sourcerange.end.column-1]
 
-        if preferred_quotation in str_contents:
+        if (preferred_quotation in str_contents and
+                not force_preferred_quotation):
             return
 
+        # Escape preferred quotes if present.
+
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: remove this blank line ;)

File: bears/general/QuotesBear.py [hunk 15]
```diff
@@ -30,13 +31,22 @@ def correct_single_line_str(self, filename, file, sourcerange,
...
+        escape_str = '\\' + preferred_quotation
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: I would still inline this one ;)

Comment (newbazz) [reply bears/general/QuotesBear.py]: What does inline mean?

File: tests/general/QuotesBearTest.py [hunk 16]
```diff
@@ -97,3 +105,34 @@ def test_invalid_quotes(self):
...
+
+        self.section['preferred_quotation'] = "'"
+        with execute_bear(self.uut, self.filename, self.no_quote_file,
+                          dependency_results=self.dep_results) as results:
+            res_list = list(results)
+            self.assertEqual(res_list[0].diffs[self.filename].unified_diff,
```
Comment (Makman2) [inline tests/general/QuotesBearTest.py]: could you also assert for the length of results and the other diff?

Comment (newbazz) [reply tests/general/QuotesBearTest.py]: other diff as in?

File: bears/general/QuotesBear.py (hunk 15 above)
Comment (newbazz) [reply bears/general/QuotesBear.py]: You mean `str_contents = str_contents.replace(preferred_quotation, '\\' + preferred_quotation)`

File: bears/general/QuotesBear.py (hunk 14 above)
Comment (newbazz) [reply bears/general/QuotesBear.py]: done

**Comment by newbazz** [comment]
@Makman2 done with the other diff.

File: bears/general/QuotesBear.py [hunk 17]
```diff
@@ -30,13 +31,21 @@ def correct_single_line_str(self, filename, file, sourcerange,
             The sourcerange indicating where to find the string.
         :param preferred_quotation:
             ``'`` or ``"`` respectively.
+        :param force_preferred_quotation:
+            Decide whether the preffered quotes are compulsory
```
Comment (Makman2) [inline bears/general/QuotesBear.py]: --> preferred

**Comment by newbazz** [comment]
@Makman2 done with typos ;P

**Comment by Makman2** [comment]
ack 68915cb

**Comment by newbazz** [comment]
how sud i now proceed @Makman2 ?

**Comment by Makman2** [comment]
with a new issue ;)

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by newbazz** [comment]
How about reviewing some issues? :P

**Comment by Makman2** [comment]
> How about reviewing some issues? :P

Sure ;)

**Comment by newbazz** [comment]
I did rebased it just a while ago!

**Comment by Makman2** [comment]
@gitmate-bot ff
